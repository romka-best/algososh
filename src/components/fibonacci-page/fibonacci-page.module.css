.root {

}

.form {
    display: flex;
    justify-content: center;
    gap: 12px;
}

.input {
    max-width: 378px;
}

.circles {
    display: flex;
    justify-content: center;
    row-gap: 48px;
    column-gap: 16px;
    flex-wrap: wrap;

    margin-top: 48px;
}
