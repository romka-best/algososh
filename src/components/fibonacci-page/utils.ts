export function* getFibonacciNumbers(number: number): Generator<Array<number>> {
    const numbers: Array<number> = [];
    for (let i = 0; i <= number; i++) {
        if (i < 2) {
            numbers.push(1);
        } else {
            numbers.push(numbers[i - 2] + numbers[i - 1]);
        }
        yield numbers;
    }

    return numbers;
}
