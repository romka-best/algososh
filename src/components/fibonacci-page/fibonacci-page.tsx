import React from 'react';

import { SHORT_DELAY_IN_MS } from '../../constants/delays';
import { ElementStates } from '../../types/element-states';

import { SolutionLayout } from '../ui/solution-layout/solution-layout';
import { Input } from '../ui/input/input';
import { Button } from '../ui/button/button';
import { Circle } from '../ui/circle/circle';

import { getFibonacciNumbers } from './utils';
import styles from './fibonacci-page.module.css';

export const FibonacciPage: React.FC = () => {
    const [ isLoading, setIsLoading ] = React.useState<boolean>(false);
    const [ isFinished, setIsFinished ] = React.useState<boolean>(false);

    const [ number, setNumber ] = React.useState<number | undefined>();
    const [ numbers, setNumbers ] = React.useState<Array<number>>([]);

    const handleButtonClick = React.useCallback((event: React.SyntheticEvent) => {
        event.preventDefault();

        if (isLoading) {
            return;
        }

        setNumbers([]);
        setIsLoading(true);
    }, [ isLoading ]);

    React.useEffect(() => {
        if (isLoading && number) {
            const generatorAlgorithm = getFibonacciNumbers(number);
            const interval = window.setInterval(() => {
                const generatorValue = generatorAlgorithm.next();

                setNumbers(() => {
                    const newState: Array<number> = [];
                    for (let i = 0; i < generatorValue.value.length; i++) {
                        newState.push(generatorValue.value[i]);
                    }
                    return newState;
                });

                if (generatorValue.done) {
                    window.clearInterval(interval);
                    setIsLoading(false);
                    setIsFinished(true);
                }
            }, SHORT_DELAY_IN_MS);
        }
    }, [ isLoading ]);

    return (
        <SolutionLayout title="Последовательность Фибоначчи">
            <div className={ styles.root }>
                <form className={ styles.form } onSubmit={ handleButtonClick }>
                    <Input
                        extraClass={ styles.input }
                        placeholder="Введите число"
                        type="number"
                        isLimitText={ true }
                        min={ 1 }
                        max={ 19 }
                        maxLength={ 2 }
                        disabled={ isLoading }
                        value={ number }
                        onChange={ (event: React.FormEvent<HTMLInputElement>) => {
                            setNumber(Number(event.currentTarget.value));
                            setIsFinished(false);
                        } }
                    />
                    <Button
                        extraClass={ styles.button }
                        text="Рассчитать"
                        disabled={ isLoading || !(number && (number >= 1 && number <= 19)) }
                        isLoader={ isLoading }
                        type="submit"
                    />
                </form>
                {
                    number && (isLoading || isFinished) ? (
                        <div className={ styles.circles }>
                            {
                                numbers.map((value, index) => (
                                        <Circle
                                            key={ index }
                                            letter={ String(value) }
                                            index={ index }
                                            state={ ElementStates.Default }
                                        />
                                    )
                                )
                            }
                        </div>
                    ) : null
                }
            </div>
        </SolutionLayout>
    );
};
