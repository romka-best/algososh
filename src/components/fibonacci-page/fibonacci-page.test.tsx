import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { FibonacciPage } from './fibonacci-page';

describe('FibonacciPage', () => {
    describe('Отображение', () => {
        it('Начальное состяние', () => {
            const page = renderer
                .create(
                    <BrowserRouter>
                        <FibonacciPage/>
                    </BrowserRouter>
                )
                .toJSON();

            expect(page).toMatchSnapshot();
        });
    });
});
