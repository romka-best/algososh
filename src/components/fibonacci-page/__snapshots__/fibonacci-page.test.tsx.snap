// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FibonacciPage Отображение Начальное состяние 1`] = `
<main
  className="content"
>
  <div
    className="titleBox"
  >
    <h1
      className="text text_type_h2 text_color_h1 title"
    >
      МБОУ АЛГОСОШ
    </h1>
    <span
      className="text text_type_fibonacci text_color_secondary subtitle"
    >
      им. Фибоначчи
    </span>
  </div>
  <div
    className="contentCard"
  >
    <a
      className="link"
      href="/"
      onClick={[Function]}
    >
      <button
        className="button returnButton"
        type="button"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M12.7073 4.29289C13.0978 4.68342 13.0978 5.31658 12.7073 5.70711L7.41436 11H19.0002C19.5524 11 20.0002 11.4477 20.0002 12C20.0002 12.5523 19.5524 13 19.0002 13H7.41436L12.7073 18.2929C13.0978 18.6834 13.0978 19.3166 12.7073 19.7071C12.3167 20.0976 11.6836 20.0976 11.293 19.7071L3.58594 12L11.293 4.29289C11.6836 3.90237 12.3167 3.90237 12.7073 4.29289Z"
            fill="#0032FF"
            fillRule="evenodd"
          />
        </svg>
        <p
          className="text text_type_button text_color_link ml-4"
        >
          К оглавлению
        </p>
      </button>
    </a>
    <h3
      className="text text_type_h3 text_color_h3 cardTitle"
    >
      Последовательность Фибоначчи
    </h3>
    <div
      className="root"
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="content input"
        >
          <input
            className="input text text_type_input text_color_input"
            disabled={false}
            max={19}
            maxLength={2}
            min={1}
            onChange={[Function]}
            placeholder="Введите число"
            type="number"
          />
          <span
            className="text text_type_input-lim text_color_input mt-2 ml-8 limit"
          >
            Максимальное число — 19
          </span>
        </div>
        <button
          className="text text_type_button text_color_primary button button"
          disabled={true}
          type="submit"
        >
          <p
            className="text"
          >
            Рассчитать
          </p>
        </button>
      </form>
    </div>
  </div>
  <p
    className="text text_type_column text_color_input mt-14 copyright"
  >
    © Сделано в Практикуме Романом Даниловым.
  </p>
</main>
`;
